sql/schemachanger: backup tests can flake due to new job detection

In https://mesolite.cluster.engflow.com/invocations/default/f31619cd-bdd1-464f-b4f3-89b303dd13fe we some times can end up with flakes in the our backup and restore tests. These are caused by the lack of precision in the finished column of the jobs table, and we need an alternate approach. We can instead take advantage of the job IDs increasing to detect the latest jobs.

These flakes take the form:

```

TestBackupSuccessMixedVersion_base_drop_column_basic/post_commit_stage_6_of_7/restore_all_tables_in_database Failure
File size: 1.8 kB
=== RUN   TestBackupSuccessMixedVersion_base_drop_column_basic/post_commit_stage_6_of_7/restore_all_tables_in_database
    backup.go:635: 
        	Error Trace:	pkg/sql/schemachanger/sctest/backup.go:635
        	Error:      	Not equal: 
        	            	expected: [][]string{[]string{"CREATE SCHEMA public"}, []string{"CREATE TABLE public.t (i INT8 NOT NULL, k INT8 NULL, CONSTRAINT t_pkey PRIMARY KEY (i ASC)) WITH ('schema_locked' = true)"}}
        	            	actual  : [][]string{[]string{"CREATE SCHEMA public"}, []string{"CREATE TABLE public.t (i INT8 NOT NULL, j INT8 NULL, k INT8 NULL, CONSTRAINT t_pkey PRIMARY KEY (i ASC)) WITH ('schema_locked' = true)"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -5,3 +5,3 @@
        	            	  ([]string) (len=1) {
        	            	-  (string) (len=121) "CREATE TABLE public.t (i INT8 NOT NULL, k INT8 NULL, CONSTRAINT t_pkey PRIMARY KEY (i ASC)) WITH ('schema_locked' = true)"
        	            	+  (string) (len=134) "CREATE TABLE public.t (i INT8 NOT NULL, j INT8 NULL, k INT8 NULL, CONSTRAINT t_pkey PRIMARY KEY (i ASC)) WITH ('schema_locked' = true)"
        	            	  }
        	Test:       	TestBackupSuccessMixedVersion_base_drop_column_basic/post_commit_stage_6_of_7/restore_all_tables_in_database
        	Messages:   	backup contents:
        	            	parent_schema_name,object_name,object_type
        	            	NULL, public, schema
        	            	public, t, table
        	            	
        	            	EXPLAIN (DDL) diagram not available: requested but never provided
--- FAIL: TestBackupSuccessMixedVersion_base_drop_column_basic/post_commit_stage_6_of_7/restore_all_tables_in_database (1.00s)
```

Jira issue: CRDB-57488

--- a/pkg/sql/schemachanger/sctest/backup.go
+++ b/pkg/sql/schemachanger/sctest/backup.go
@@ -15,7 +15,6 @@ import (
 	"strings"
 	"sync/atomic"
 	"testing"
-	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/clusterversion"
 	"github.com/cockroachdb/cockroach/pkg/sql/parser"
@@ -422,10 +421,10 @@ func backupRollbacks(t *testing.T, factory TestServerFactory, cs CumulativeTestC
 		// Fetch the state of the cluster before the schema change kicks off.
 		tdb.Exec(t, fmt.Sprintf("USE %q", cs.DatabaseName))
 		expected := parserRoundTrip(t, tdb.QueryStr(t, fetchDescriptorStateQuery))
-		// Store the start time, so we know which jobs are relevant.
-		ts := tdb.QueryRow(t, "SELECT current_timestamp(0)::timestamp")
-		var startTime time.Time
-		ts.Scan(&startTime)
+		// Store the highest job ID, so we know which jobs are relevant.
+		jobID := tdb.QueryRow(t, "SELECT COALESCE(max(job_id), 0) FROM [SHOW JOBS]")
+		var maxJobID int64
+		jobID.Scan(&maxJobID)
 		// Kick off the schema change, fail it at the prescribed stage,
 		// and perform the backups during the rollback via the BeforeStage hook.
 		require.Regexp(t, fmt.Sprintf("boom %d", cs.StageOrdinal),
@@ -437,7 +436,7 @@ func backupRollbacks(t *testing.T, factory TestServerFactory, cs CumulativeTestC
 		// Note: Depending on the state of the schema objects, no job may be created
 		// to finalize the rollback, since dependent objects were never backed
 		// up.
-		succeeded, jobExists := hasLatestSchemaChangeSucceededWithTimestamp(t, tdb, startTime)
+		succeeded, jobExists := hasLatestSchemaChangeSucceededWithMaxJobID(t, tdb, maxJobID)
 		require.False(t, succeeded && jobExists)
 		tdb.Exec(t, fmt.Sprintf("USE %q", cs.DatabaseName))
 		postRollback := parserRoundTrip(t, tdb.QueryStr(t, fetchDescriptorStateQuery))
@@ -579,10 +578,11 @@ func exerciseBackupRestore(
 				tdb.Exec(t, cs.CreateDatabaseStmt)
 			}
 
-			// Store the start time, so we know which jobs are relevant.
-			ts := tdb.QueryRow(t, "SELECT current_timestamp(0)::timestamp")
-			var startTime time.Time
-			ts.Scan(&startTime)
+			// Store the highest job ID, so we know which jobs are relevant.
+			tdb.Exec(t, "USE system")
+			jobID := tdb.QueryRow(t, "SELECT COALESCE(max(job_id), 0) FROM [SHOW JOBS]")
+			var maxJobID int64
+			jobID.Scan(&maxJobID)
 
 			// RESTORE.
 			pe.RequestPlan()
@@ -598,7 +598,7 @@ func exerciseBackupRestore(
 			skipComparison := false
 			// If the only thing left is to clean up then this restore may have no
 			// job so the startTime is important.
-			if succeeded, jobExists := hasLatestSchemaChangeSucceededWithTimestamp(t, tdb, startTime); (succeeded && jobExists) || (b.maySucceed && !jobExists) {
+			if succeeded, jobExists := hasLatestSchemaChangeSucceededWithMaxJobID(t, tdb, maxJobID); (succeeded && jobExists) || (b.maySucceed && !jobExists) {
 				expected = b.expectedOnSuccess
 				skipComparison = b.skipComparisonOnSuccess
 				if !b.maySucceed {

--- a/pkg/sql/schemachanger/sctest/framework.go
+++ b/pkg/sql/schemachanger/sctest/framework.go
@@ -1015,14 +1015,19 @@ func waitForSchemaChangesToFinish(t *testing.T, tdb *sqlutils.SQLRunner) {
 	)
 }
 
-func hasLatestSchemaChangeSucceededWithTimestamp(
-	t *testing.T, tdb *sqlutils.SQLRunner, startTime time.Time,
+// hasLatestSchemaChangeSucceededWithMaxJobID detects if the latest schema change
+// has changed. The function requires the last observed maximum job ID and takes
+// advantage of the increasing nature of the ID (since the upper bits encode
+// the current time). Note: We could have used `finished`, but it does not encode
+// enough precision.
+func hasLatestSchemaChangeSucceededWithMaxJobID(
+	t *testing.T, tdb *sqlutils.SQLRunner, maxJobID int64,
 ) (succeeded bool, jobExists bool) {
 	result := tdb.QueryStr(t, fmt.Sprintf(
-		`SELECT status FROM [SHOW JOBS] WHERE job_type IN ('%s') AND finished >= $1::timestamp ORDER BY finished DESC, job_id DESC LIMIT 1`,
+		`SELECT status FROM [SHOW JOBS] WHERE job_type IN ('%s') AND job_id > $1 ORDER BY finished DESC, job_id DESC LIMIT 1`,
 		jobspb.TypeNewSchemaChange,
 	),
-		startTime)
+		maxJobID)
 	return len(result) == 0 || result[0][0] == "succeeded", len(result) > 0
 }
 

